common/lib/plugins/failover2/failover2_plugin.ts

```diff
@@ -0,0 +1,502 @@
+    if (props.has(Failover2Plugin.INTERNAL_CONNECT_PROPERTY_NAME)) {
+      return await this._staleDnsHelper.getVerifiedConnection(hostInfo.host, isInitialConnection, this.hostListProviderService!, props, connectFunc);
+    }
+
+    let client: ClientWrapper = null;
+
+    if (!this.enableFailoverSetting || !WrapperProperties.ENABLE_CLUSTER_AWARE_FAILOVER.get(props)) {
+      return await this._staleDnsHelper.getVerifiedConnection(hostInfo.host, isInitialConnection, this.hostListProviderService!, props, connectFunc);
+    }
+
+    const hostInfoWithAvailability: HostInfo = this.pluginService.getHosts().find((x) => x.getHostAndPort() === hostInfo.getHostAndPort());
+
+    if (!hostInfoWithAvailability || hostInfoWithAvailability.getAvailability() != HostAvailability.NOT_AVAILABLE) {
+      try {
+        client = await this._staleDnsHelper.getVerifiedConnection(
+          hostInfo.host,
+          isInitialConnection,
+          this.hostListProviderService!,
+          props,
+          connectFunc
+        );
+      } catch (error) {
+        if (!this.shouldErrorTriggerClientSwitch(error)) {
+          throw error;
+        }
+
+        this.pluginService.setAvailability(hostInfo.allAliases, HostAvailability.NOT_AVAILABLE);
+
+        try {
+          await this.failover(hostInfo);
+        } catch (error) {
+          if (error instanceof FailoverSuccessError) {
+            client = this.pluginService.getCurrentClient().targetClient;
+          } else {
+            throw error;
+          }
+        }
+      }
+    } else {
+      try {
+        await this.pluginService.refreshHostList();
+        await this.failover(hostInfo);
+      } catch (error) {
+        if (error instanceof FailoverSuccessError) {
+          client = this.pluginService.getCurrentClient().targetClient;
+        } else {
+          throw error;
+        }
+      }
+    }
+
+    if (!client) {
+      // This should be unreachable, the above logic will either get a connection successfully or throw an exception.
+      throw new AwsWrapperError(Messages.get("Failover2.unableToConnect"));
+    }
+
+    if (isInitialConnection) {
+      await this.pluginService.refreshHostList(client);
+    }
+
+    return client;
+  }
+
+  override async execute<T>(methodName: string, methodFunc: () => Promise<T>): Promise<T> {
+    try {
+      // Verify there aren't any unexpected error emitted while the connection was idle.
+      if (this.pluginService.hasNetworkError()) {
+        // Throw the unexpected error directly to be handled.
+        throw this.pluginService.getUnexpectedError();
+      }
+
+      if (!this.enableFailoverSetting || this.canDirectExecute(methodName)) {
+        return await methodFunc();
+      }
+
+      if (this.isClosed) {
+        await this.invalidInvocationOnClosedConnection();
+      }
+
+      if (this.canUpdateTopology(methodName)) {
+        await this.updateTopology();
+      }
```

Comment:
Could you please verify with @sergiyvamz if this is necessary? JDBC's implementation doesn't have this check.